providers/autodns/api.go

```diff
@@ -0,0 +1,143 @@
+package autodns
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sort"
+
+	"github.com/StackExchange/dnscontrol/v3/models"
+)
+
+type ZoneListFilter struct {
+	Key string `json:"key"`
+	Value string `json:"value"`
+	Operator string `json:"operator"`
+	Link string `json:"link,omitempty"`
+	Filter []*ZoneListFilter `json:"filters,omitempty"`
+}
+
+type ZoneListRequest struct {
+	Filter []*ZoneListFilter `json:"filters"`
+}
+
+func (api *autoDnsProvider) request(method string, requestPath string, data interface{}) ([]byte, error) {
+	client := &http.Client{}
+
+	requestUrl := api.baseURL
+	requestUrl.Path = api.baseURL.Path + requestPath
+
+	request := &http.Request{
+		URL: &requestUrl,
+		Header: api.defaultHeaders,
+		Method: method,
+	}
+
+	if data != nil {
+		body, _ := json.Marshal(data)
+
+		buffer := bytes.NewBuffer(body)
+
+		request.Body = io.NopCloser(buffer)
+	}
+
+	response, error := client.Do(request)
+
+	if error != nil {
+		return nil, error
+	}
+
+	defer response.Body.Close()
+
+	responseText, _ := ioutil.ReadAll(response.Body)
+
+	if response.StatusCode != 200 {
+		return nil, errors.New("Request to " + requestUrl.Path + " failed: " + string(responseText))
+	}
+
+	return responseText, nil
+}
+
+func (api *autoDnsProvider) findZoneSystemNameServer(domain string) (*models.Nameserver, error) {
+	request := &ZoneListRequest{}
+
+	request.Filter = append(request.Filter, &ZoneListFilter{
+		Key:      "name",
+		Value:    domain,
+		Operator: "EQUAL",
+	})
+
+	responseData, err := api.request("POST", "zone/_search", request)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var responseObject JSONResponseDataZone
+	_ = json.Unmarshal(responseData, &responseObject)
+
+	if len(responseObject.Data) != 1 {
+		return nil, errors.New("Domain " + domain + " could not be found in AutoDNS")
+	}
+
+	systemNameServer := &models.Nameserver{Name: responseObject.Data[0].SystemNameServer}
+
+	return systemNameServer, nil
+}
+
+func (api *autoDnsProvider) getZone(domain string) (*Zone, error) {
+	systemNameServer, err := api.findZoneSystemNameServer(domain)
+
```

Comment:
style: remove blank between verb and error checking.

Reply:
Hopefully I fixed this with https://github.com/StackExchange/dnscontrol/pull/1336/commits/50697d7dd6a45eace6e33de2f1678daf828b44bd

I did a bit of both, the code was not very clear at first. The response check however can be seen as unnecessary because the zone is already found in the system by the previous call.

I've added checking of a JSON decode (_unmarshal_) error since the returned data format might be changed.